### PR TITLE
Fix services cancellation

### LINKED ISSUE #1293: Services don't cancel immediately if user presses on stop
Current implementation don't allow services to stop immediately as user presses on stop, they wait for next file/entry to recheck if user has cancelled it, leaving the current execution at it's progress.
What makes this worse is, user is not notified, instead notifications cancel immediately leaving user thinking the progress has stopped. 
Effects all services implementation.

### PR DESCRIPTION
**Pull request by VishalNehra**
Fixes #1293

### CODE REVIEW

**Comment by VishalNehra** [comment]
@EmmanuelMess please review,, we've the release on hold.

**Comment by EmmanuelMess** [comment]
Compress cancellation leaves the notification up and progress doesn't stop. But compressed file is never created (tried twice).
Extraction, copy and cut cancellation work correctly.
Can't encrypt zip file (50MB), can't test encryption or decryption. Doesn't show an error, notfication doesn't show.

**Comment by VishalNehra** [comment]
Noted. Will work on a fix.

**Comment by VishalNehra** [comment]
Fixed cancellation for compression, as for encrypt/decrypt they seem to be working fine.

**Comment by EmmanuelMess** [comment]
Okay, if I stop the compression on the "undefined flickering state" before it actually starts the app crashes:
```
    --------- beginning of crash
06-22 08:57:03.120 com.amaze.filemanager E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #4
    Process: com.amaze.filemanager, PID: 10103
    java.lang.RuntimeException: An error occurred while executing doInBackground()
        at android.os.AsyncTask$3.done(AsyncTask.java:309)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
        at java.util.concurrent.FutureTask.run(FutureTask.java:242)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:234)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
        at java.lang.Thread.run(Thread.java:818)
     Caused by: java.lang.IllegalStateException: This is not the first datapoint!
        at com.amaze.filemanager.asynchronous.services.AbstractProgressiveService.addFirstDatapoint(AbstractProgressiveService.java:213)
        at com.amaze.filemanager.asynchronous.services.ZipService$CompressAsyncTask.doInBackground(ZipService.java:228)
        at com.amaze.filemanager.asynchronous.services.ZipService$CompressAsyncTask.doInBackground(ZipService.java:199)
        at android.os.AsyncTask$2.call(AsyncTask.java:295)
        at java.util.concurrent.FutureTask.run(FutureTask.java:237)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:234) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588) 
        at java.lang.Thread.run(Thread.java:818) 
06-22 08:57:03.133 ? W/ActivityManager:   Force finishing activity com.amaze.filemanager/.activities.MainActivity
06-22 08:57:03.505 com.amaze.filemanager E/Surface: getSlotFromBufferLocked: unknown buffer: 0xb91aa520
```
And I have a partial zip file.

If I stop after the compression starts, it stops, but I still have the partial file.

**Comment by EmmanuelMess** [comment]
Stopping encryption also leaves me with an incomplete encryption folder. Trying to encrypt something I can cancel the decryption of crashes or doesn't start (see #1309).

**Comment by TranceLove** [comment]
The first datapoint problem seems to be the most teething problem we have with - also seen in #1222.

Seriously, I'd recommend you to write an Espresso test for verifying correctness of code if you can. Robolectric is not enough, since our workstations are (usually) much more powerful than Android devices and may help code to avoid some bottlenecks.

**Comment by EmmanuelMess** [comment]
I think most are concurrency issues, can Espresso deal with multiple threads and test ~1000 times?

**Comment by TranceLove** [comment]
> I think most are concurrency issues, can Espresso deal with multiple threads?

Not sure if [Multiprocess Espresso](https://developer.android.com/training/testing/espresso/multiprocess) is what you're talking about. Otherwise, I think the limit is the device's limit.

> and test ~1000 times?

Does it mean to loop the test 1000 times?

**Comment by EmmanuelMess** [comment]
> Does it mean to loop the test 1000 times?

Generally concurrency issues arise probabilistically, that's why. Test 1000 time, tell me how many failed and why.

**Comment by VishalNehra** [comment]
Firstly, I can't reproduce the *flickering* thing. Notification is build only once and then it's being notified. If the notify call is the cause of flicker then I don't think it can be helped.
Tell me your configuration so that I can test.
As for decryption cancellation, the limitation is from the java API. The helper classes load the whole file in a single buffer internally. So it's either completely decrypted or not at all. That's why they leave a 0 mb file on cancellation. That's also the reason why we can't see decryption progress until it's fully completed.
As for the datapoint problem, I think I did find out the cause for it earlier and had plans to fix it. But debugging it again all over is gonna take some time. Given it's really hard to reproduce in ideal conditions. 
Let me know your configs so I can test both datapoint and notification flicker. 
I honestly think both these issues are out of bounds of this PR and shall be assigned their separate issues.

**Comment by EmmanuelMess** [comment]
My device is Moto E Gen 2 4G

**Comment by EmmanuelMess** [comment]
>  That's why they leave a 0 mb file on cancellation. 

Why not delete the file?

**Comment by VishalNehra** [comment]
Not necessary. We don't also delete in other services afterall.

**Comment by EmmanuelMess** [comment]
> Not necessary. We don't also delete in other services afterall.

Other services don't leave a 0B file on cancellation. How it not necessary for encryption?

**Comment by VishalNehra** [comment]
No temporary file creation. The bytes are written directly at the target in all these functions.

**Comment by EmmanuelMess** [comment]
> No temporary file creation. The bytes are written directly at the target in all these functions.

No file was there after cancellation in the case of compression (for example).

**Comment by EmmanuelMess** [comment]
Also, the point still stands, the user deletes the file via the cancellation of its creation.

**Comment by VishalNehra** [comment]
Cancel is meant to cancel the process at whatever point it was in at that instance. It is not an 'undo' option.

**Comment by EmmanuelMess** [comment]
> Cancel is meant to cancel the process at whatever point it was in at that instance. It is not an 'undo' option.

I propose it is, definition of cancel:

> to decide that an organized event will not happen, or to stop an order for goods or services that you no longer want

-- [source](https://dictionary.cambridge.org/es/diccionario/ingles/cancel)

The creation of the file should no longer happen, the user no longer wants it. Why would the user want "half" a zip file?

**Comment by VishalNehra** [comment]
Please create a new issue and assign it to me. Will be working on it for new milestone.

**Comment by EmmanuelMess** [comment]
Could you test flickering?

**Comment by VishalNehra** [comment]
Fixed #1305 (only the left margin, can't fix framework layouts, not in my hand)
Also please check and confirm if flickering still exists.

**Comment by EmmanuelMess** [comment]
>  Also please check and confirm if flickering still exists.
Yes, will post a video ASAP

File: app/src/main/java/com/amaze/filemanager/ui/views/drawer/CustomNavigationView.java [hunk 1]
```diff
@@ -66,7 +67,7 @@ public void deselectItems() {
 
     @Override
     public Parcelable onSaveInstanceState() {
-        if (!doesNavigationViewSavedStateExist()) {
```
[the remark this answers is not in the document]
Comment (VishalNehra) [reply app/src/main/java/com/amaze/filemanager/ui/views/drawer/CustomNavigationView.java]: Modified that code by mistake. Was testing something. Reverted.

[the remark this answers is not in the document]
Comment (EmmanuelMess) [reply app/src/main/java/com/amaze/filemanager/ui/views/drawer/CustomNavigationView.java]: You reverted half the changes only, `SavedState` still extends `NavigationView.SavedState` and `SavedState(Parcel)` still calls  `super(Parcel, ClassLoader)`. I think you are overextending the scope of the PR with that.

**Comment by EmmanuelMess** [comment]
> Fixed #1305 (only the left margin, can't fix framework layouts, not in my hand)

This extends the scope of PR, please don't do that.

**Comment by VishalNehra** [comment]
> This extends the scope of PR, please don't do that.
That margin was put by mistake. I'm in no way trying to solve #1305

File: app/src/main/java/com/amaze/filemanager/ui/views/drawer/CustomNavigationView.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (VishalNehra) [reply app/src/main/java/com/amaze/filemanager/ui/views/drawer/CustomNavigationView.java]: Like I said, I was trying something else... will revert these changes too. Should've reviewed before making a commit :(

**Comment by VishalNehra** [comment]
Done reverting back the drawer class.

**Pull request merged**